File: app/dtos/notifications.py
from datetime import datetime, time
from enum import StrEnum

from pydantic import BaseModel, model_validator


class NotificationType(StrEnum):
    GENERAL = "GENERAL"
    PREDICTION = "PREDICTION"
    CHALLENGE = "CHALLENGE"
    ADVICE = "ADVICE"
    REPORT = "REPORT"


class NotificationResponse(BaseModel):
    notification_id: int
    notification_type: NotificationType
    title: str
    message: str
    link_url: str | None = None
    is_read: bool
    read_at: datetime | None = None
    created_at: datetime


class NotificationUnreadCountResponse(BaseModel):
    unread_count: int


class NotificationMarkReadResponse(BaseModel):
    notification_id: int
    is_read: bool
    read_at: datetime


class NotificationMarkAllReadResponse(BaseModel):
    updated_count: int
    unread_count: int = 0


class NotificationPreferenceResponse(BaseModel):
    push_enabled: bool
    health_data_reminder_enabled: bool
    challenge_mission_enabled: bool
    prediction_result_enabled: bool
    advice_update_enabled: bool
    virtual_pet_enabled: bool
    email_enabled: bool
    weekly_report_enabled: bool
    important_notice_enabled: bool
    promotion_enabled: bool
    quiet_start_time: time
    quiet_end_time: time


class NotificationPreferenceUpdateRequest(BaseModel):
    push_enabled: bool | None = None
    health_data_reminder_enabled: bool | None = None
    challenge_mission_enabled: bool | None = None
    prediction_result_enabled: bool | None = None
    advice_update_enabled: bool | None = None
    virtual_pet_enabled: bool | None = None
    email_enabled: bool | None = None
    weekly_report_enabled: bool | None = None
    important_notice_enabled: bool | None = None
    promotion_enabled: bool | None = None
    quiet_start_time: time | None = None
    quiet_end_time: time | None = None

    @model_validator(mode="after")
    def validate_quiet_times(self):
        if self.quiet_start_time and self.quiet_end_time and self.quiet_start_time == self.quiet_end_time:
            raise ValueError("알림 시작 시간과 종료 시간은 달라야 합니다.")
        return self

File: frontend/src/api/notifications.ts
import { apiRequest } from "./client";

export type NotificationItem = {
  notification_id: number;
  notification_type: "GENERAL" | "PREDICTION" | "CHALLENGE" | "ADVICE" | "REPORT";
  title: string;
  message: string;
  link_url: string | null;
  is_read: boolean;
  read_at?: string | null;
  created_at: string;
};

export type NotificationPreference = {
  push_enabled: boolean;
  health_data_reminder_enabled: boolean;
  challenge_mission_enabled: boolean;
  prediction_result_enabled: boolean;
  advice_update_enabled: boolean;
  virtual_pet_enabled: boolean;
  email_enabled: boolean;
  weekly_report_enabled: boolean;
  important_notice_enabled: boolean;
  promotion_enabled: boolean;
  quiet_start_time: string;
  quiet_end_time: string;
};

export async function getUnreadNotificationCount(token?: string) {
  return apiRequest<{ data: { unread_count: number } }>("/notifications/unread-count", { token });
}

export async function getNotifications(token?: string) {
  return apiRequest<{ data: NotificationItem[] }>("/notifications", { token });
}

export async function markNotificationRead(notificationId: number, token?: string) {
  return apiRequest<{ data: { notification_id: number; is_read: boolean; read_at: string } }>(
    `/notifications/${notificationId}/read`,
    {
      method: "PATCH",
      token,
    },
  );
}

export async function markAllNotificationsRead(token?: string) {
  return apiRequest<{ data: { updated_count: number; unread_count: number } }>("/notifications/read-all", {
    method: "PATCH",
    token,
  });
}

export async function getNotificationPreferences(token?: string) {
  return apiRequest<{ data: NotificationPreference }>("/notification-preferences", { token });
}

export async function updateNotificationPreferences(payload: Partial<NotificationPreference>, token?: string) {
  return apiRequest<{ data: NotificationPreference }>("/notification-preferences", {
    method: "PATCH",
    body: JSON.stringify(payload),
    token,
  });
}
